extension/function.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extension
+
+import (
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// FunctionContext is a interface to provide context to the custom function
+type FunctionContext interface {
+	EvalArgs(row chunk.Row) ([]types.Datum, error)
+}
+
+// FunctionDef is the definition for the custom function
+type FunctionDef struct {
+	Name   string
+	EvalTp types.EvalType
+	ArgTps []types.EvalType
+	// EvalStringFunc is the eval function when `EvalTp` is `types.ETString`
+	EvalStringFunc func(ctx FunctionContext, row chunk.Row) (string, bool, error)
+	// EvalIntFunc is the eval function when `EvalTp` is `types.ETInt`
+	EvalIntFunc func(ctx FunctionContext, row chunk.Row) (int64, bool, error)
```

Comment:
As an interface, for now we only have `EvalString`/`EvalInt` supported and I was a little concern that if we want to provide more types support in the future, all the existing extensions should be updated accordingly.

However, this might not be a big issue... I don't have a strong opinion.

Reply:
I think in the future if we have more requirements, we just need to add `EvalXXXFunc`. The old extensions will not be updated because it is only a struct, the `EvalXXXFunc` will default be nil